### PR TITLE
Unify shard failure handling in getIterators

### PR DESCRIPTION
**Pull request by mfussenegger**
Shard-failure exception handling when retrieving a `BatchIterator` was
split across three places:

- `ShardCollectSource#getIterators`
- `ShardCollectorProvider.getFutureIterator`
- `ShardCollectorProvider.getFutureOrderedCollector`

This moves the exception handling from `ShardCollectorProvider` to the
`ShardCollectSource`, to have everything in a single place.

### CODE REVIEW

File: server/src/main/java/io/crate/execution/engine/collect/ShardCollectorProvider.java [hunk 1]
```diff
@@ -102,63 +100,62 @@ public abstract class ShardCollectorProvider {
...
-                futureIt.complete(getIterator(collectPhase, requiresScroll, collectTask));
-            } catch (ShardNotFoundException | IllegalIndexShardStateException e) {
-                if (IndexParts.isPartitioned(e.getIndex().getName())) {
-                    futureIt.complete(InMemoryBatchIterator.empty(SentinelRow.SENTINEL));
-                } else {
-                    futureIt.completeExceptionally(e);
+    public class BiFactory {
+
+        public BatchIterator<Row> getIterator(RoutedCollectPhase collectPhase,
+                                            boolean requiresScroll,
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/collect/ShardCollectorProvider.java:109]: minor: indentation.

File: server/src/main/java/io/crate/execution/engine/collect/ShardCollectorProvider.java [hunk 2]
```diff
@@ -102,63 +100,62 @@ public abstract class ShardCollectorProvider {
...
-        if (isOpenIndex && WhereClause.canMatch(normalizedCollectNode.where())) {
-            iterator = getUnorderedIterator(normalizedCollectNode, requiresScroll, collectTask);
-        } else {
-            iterator = InMemoryBatchIterator.empty(SentinelRow.SENTINEL);
+
+        public OrderedDocCollector getOrderedCollector(RoutedCollectPhase collectPhase,
+                                                       SharedShardContext sharedShardContext,
+                                                       CollectTask collectTask,
+                                                       boolean requiresRepeat) {
+            return ShardCollectorProvider.this.getOrderedCollector(
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/engine/collect/ShardCollectorProvider.java:145]: I am noot sure how that would work given the inheritance? 

I introduced the inner `BatchIteratorFactory` mostly to enforce that users have to go through `awaitShardSearchActive` to gain access to the methods. `getOrderedCollector` is an abstract method.

**Comment by mfussenegger** [comment]
Rebased this, after https://github.com/crate/crate/pull/13828 the flakyness should be fixed.
I'm not entirely sure if it's worth changing.

**Comment by matriv** [comment]
> I'm not entirely sure if it's worth changing.

I'm not very sure as well, I think I slightly prefer this new approach, maybe ask for some other opinion?

**Review by matriv — APPROVED** [review]

**Pull request merged**